subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,542 @@
+
+static const char smp_images_str[] = "images";
+#define IMAGES_STR_LEN (sizeof(smp_images_str) - 1)
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size)
+{
+	for (int i = 0; i < map_size; i++) {
+		/* Clear found status */
+		map[i].found = false;
+	}
+}
+
+static int image_state_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_image_gr_rsp response = {0};
+	zcbor_state_t zsd[CBOR_ENCODER_STATE_NUM + 2];
+	struct zcbor_string value = {0};
+	int slot_num, img_num, rc;
+	struct zcbor_string hash, version;
+	bool bootable, pending, confirmed, active, permanent, ok;
+	size_t decoded;
+	struct zcbor_map_decode_key_val list_res_decode[] = {
+		/** Mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("version", zcbor_tstr_decode, &version),
+		ZCBOR_MAP_DECODE_KEY_DECODER("hash", zcbor_bstr_decode, &hash),
+		ZCBOR_MAP_DECODE_KEY_DECODER("slot", zcbor_int32_decode, &slot_num),
+		/** End of mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("image", zcbor_int32_decode, &img_num),
+		ZCBOR_MAP_DECODE_KEY_DECODER("bootable", zcbor_bool_decode, &bootable),
+		ZCBOR_MAP_DECODE_KEY_DECODER("pending", zcbor_bool_decode, &pending),
+		ZCBOR_MAP_DECODE_KEY_DECODER("confirmed", zcbor_bool_decode, &confirmed),
+		ZCBOR_MAP_DECODE_KEY_DECODER("active", zcbor_bool_decode, &active),
+		ZCBOR_MAP_DECODE_KEY_DECODER("permanent", zcbor_bool_decode, &permanent)};
+
+	response.cmd = IMG_MGMT_ID_STATE;
+	response.image_state.image_list_length = 0;
+	response.image_state.image_list = NULL;
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		rc = 0;
+		goto out;
+	}
+
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	ok = zcbor_map_start_decode(zsd);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+
+	ok = zcbor_tstr_decode(zsd, &value);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+	if (value.len != IMAGES_STR_LEN || memcmp(value.value, smp_images_str, IMAGES_STR_LEN)) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+
+	ok = zcbor_list_start_decode(zsd);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+
+	image_count = 0;
+
+	for (int slot = 0; slot < MCUMGR_IMAGE_LIST_SIZE; slot++) {
+		decoded = 0;
+		reset_image_map_decoder_list(list_res_decode, ARRAY_SIZE(list_res_decode));
+		/* Init buffer values */
+		active = false;
+		bootable = false;
+		confirmed = false;
+		pending = false;
+		permanent = false;
+		img_num = 0;
+		slot_num = -1;
+		hash.len = 0;
+		version.len = 0;
+
+		rc = zcbor_map_decode_bulk(zsd, list_res_decode, ARRAY_SIZE(list_res_decode),
+					   &decoded);
+		if (rc) {
+			if (slot) {
+				/* No more map */
+				break;
+			}
+			LOG_ERR("Corrupted Image data %d", rc);
+			response.status = MGMT_ERR_ECORRUPT;
+			rc = MGMT_ERR_EINVAL;
```

Comment:
Is it corrupt or an invalid value?

Reply:
I start Think here that it should use MGMT_ERR_EINVAL to status also. Is it make sense @nordicjm ?